### PR TITLE
Reduce metadata calls

### PR DESCRIPTION
**Pull request by ccleung**
Tracker Ticket: 
https://quandl.atlassian.net/browse/AP-1631

## Description:
* Minimize Dataset data fetched from server when performing a `MergedDataset([...]).data()` call:
 * If only one column index is requested, pass as query param `column_index=x` so that the server can give us back that one column
 * For more than 1 column specified, fetch all the data from the server, and filter out on the client side the columns the user has requested
   * NOTE: We have to handle multiple column index per Dataset in a MergedDataset on the client side because the API does not support multiple column_indexes for a Dataset data fetch
* Don't make metadata calls when calling `MergedDataset('[]').data()`
* Make metadata call only when user asks for an attribute of the `MergedDataset`, e.g., `MergedDataset([...]).name`
* MergedDataList (which is the class that holds the dataset data response data for a MergedDataset):
 * instead of getting the `start_date` and `end_date` from the metadata, just set these values to what we see in the actual data we got back via the merged dataframe
* Validations:
 * Do not perform client side validation of requested column indexes when performing MergedDataset().data()
 * Perform the validation after we retrieve the data
 * This prevents the forced
* Analyst `Quandl.get([....])`
 * Fail silently when making single dataset calls where the column does not exist (i.e., insert a Not Found empty column in the data frame)
 * Fail silently when making multiple dataset calls where the column does not exist
 * Fail silently when dataset does not exist ONLY in multiple dataset scenario
 * Raise a `NotFoundError` for single dataset call and dataset does not exist
 * Fail silently means to insert an empty DataFrame column whenever column(s) fails to be retrieved
* Developer flow MergedDataset and Dataset:
 * Raise a `ColumnNotFound` when dataset column does not exist
 * Raise a `NotFoundError` (as before) when dataset does not exist

## Note to QA:
* Lots can break from this PR

### CODE REVIEW

File: quandl/model/data_mixin.py [hunk 1]
```diff
@@ -35,3 +37,10 @@ def to_numpy(self):
 
     def to_csv(self):
         return self.to_pandas().to_csv()
+
+    def validate_col_index(self, df, keep_column_indexes):
```
Comment (mbasset) [inline quandl/model/data_mixin.py]: Possibly make this private method?

File: quandl/model/data_mixin.py [hunk 2]
```diff
@@ -35,3 +37,10 @@ def to_numpy(self):
...
+        num_columns = len(df.columns)
+        for col_index in keep_column_indexes:
+            if col_index > num_columns or col_index < 1:
+                raise ColumnNotFound('Requested column index %s does not exist!'
```
Comment (mbasset) [inline quandl/model/data_mixin.py]: Please check with Carrie on the translation for this.

File: quandl/model/merged_dataset.py [hunk 3]
```diff
@@ -141,30 +158,33 @@ def _initialize_raw_data(self, datasets):
...
-        # if column_indexes are specified, check that they actually exist
-        max_column_index = len(dataset.column_names) - 1
-        for index in dataset.column_index:
-            if index > max_column_index or index < 1:
-                raise ValueError("%s : requested index %s is out of range. \
-                                  Min index is 1 and Max index is %s" % (code, index,
-                                                                         max_column_index))
-        return dataset
+                    "%s : column_index needs to be a list of integer indexes" % code_str)
+            return column_index
```
Comment (mbasset) [inline quandl/model/merged_dataset.py]: Does this return a number or a list of numbers. If it's a single number it's a bit odd the default response is a list. If its a list no problem.

File: quandl/model/data_mixin.py (hunk 2 above)
Comment (ccleung) [inline quandl/model/data_mixin.py]: from carrier: "It's fine as is. I would just remove the !" I'll remove the exclamation mark

File: quandl/model/merged_dataset.py (hunk 3 above)
Comment (ccleung) [inline quandl/model/merged_dataset.py]: it's always a list of numbers.

File: quandl/model/data_mixin.py (hunk 1 above)
Comment (ccleung) [inline quandl/model/data_mixin.py]: updated

File: quandl/model/merged_dataset.py (hunk 3 above)
Comment (mbasset) [inline quandl/model/merged_dataset.py]: cool

**Comment by mbasset** [comment]
:+1:

**Pull request merged**